### PR TITLE
[SimpleCPUOffloadConnector]: Support cross layer KV layout

### PR DESCRIPTION
**Pull request by jonathanc-n**
## Purpose

Support cross layer kv cache layout. Changes are quite simple since the group is guaranteed to be produced from the same backend.

Tests were generated by Claude, and reviewed by me.

### CODE REVIEW

File: vllm/v1/simple_kv_offload/worker.py [hunk 1]
```diff
@@ -181,6 +182,74 @@ def _repr_tensor(v: torch.Tensor | list[torch.Tensor]) -> torch.Tensor:
...
+
+        # View the raw storage as (num_blocks, page_size_bytes)
+        storage = kv_cache.untyped_storage()
+        assert len(storage) % page_size_bytes == 0
+        num_blocks = len(storage) // page_size_bytes
+        gpu_tensor = (
+            torch.tensor([], dtype=torch.int8, device=self.device)
+            .set_(storage)
+            .view(num_blocks, page_size_bytes)
+        )
```
[the remark this answers is not in the document]
Comment (jonathanc-n) [reply vllm/v1/simple_kv_offload/worker.py:208]: Cannot be `UniformKVCacheSpecs`, cross layer is guarded by asserting that it is an `AttentionSpec`

**Review by NickLucche — COMMENTED** [review]
Thanks for the work @jonathanc-n !
Changes look fine to me. 
Would it be possible to run a quick benchmark showing cross-layer kv cache perf advantage here?

cc @orozery

**Comment by jonathanc-n** [comment]
Ran benchmark on L4 with:
```
CUDA_VISIBLE_DEVICES=0 .venv/bin/python benchmarks/benchmark_cross_layer_kv_dma.py \
  --num-layers 32 \
  --num-blocks 512 \
  --transfer-blocks 1 2 4 8 16 32 64 128 \
  --warmup-iters 20 \
  --iters 100 \
```

Full fixed benchmark can be found in [this commit](b27debc3ef28f2513bca9213593c215836e3001a) if want to try locally. 

Results: 
```
Speedup: per_layer_mean_ms / cross_layer_mean_ms
load  blocks=1    speedup=1.33x descriptor_reduction=32.0x
load  blocks=2    speedup=1.30x descriptor_reduction=32.0x
load  blocks=4    speedup=1.28x descriptor_reduction=32.0x
load  blocks=8    speedup=1.26x descriptor_reduction=32.0x
load  blocks=16   speedup=1.26x descriptor_reduction=32.0x
load  blocks=32   speedup=1.26x descriptor_reduction=32.0x
load  blocks=64   speedup=1.25x descriptor_reduction=32.0x
load  blocks=128  speedup=1.25x descriptor_reduction=32.0x
store blocks=1    speedup=1.32x descriptor_reduction=32.0x
store blocks=2    speedup=1.25x descriptor_reduction=32.0x
store blocks=4    speedup=1.21x descriptor_reduction=32.0x
store blocks=8    speedup=1.20x descriptor_reduction=32.0x
store blocks=16   speedup=1.20x descriptor_reduction=32.0x
store blocks=32   speedup=1.24x descriptor_reduction=32.0x
store blocks=64   speedup=1.23x descriptor_reduction=32.0x
store blocks=128  speedup=1.23x descriptor_reduction=32.0x
```

**Review by NickLucche — APPROVED** [review]
Thanks @jonathanc-n , this works for me! 
Can we also get a quick second look at it from @ivanium

File: vllm/v1/simple_kv_offload/worker.py [hunk 2]
```diff
@@ -181,6 +183,74 @@ def _repr_tensor(v: torch.Tensor | list[torch.Tensor]) -> torch.Tensor:
             self.store_stream,
         )
 
+    def register_cross_layers_kv_cache(
+        self,
+        kv_cache: torch.Tensor,
+        attn_backend: type["AttentionBackend"],
+    ) -> None:
```
Comment (ivanium) [inline vllm/v1/simple_kv_offload/worker.py:193]: Can we just reuse register_kv_caches()?
```suggestion
    def register_cross_layers_kv_cache(
        self,
        kv_cache: torch.Tensor,
        attn_backend: type["AttentionBackend"],
    ) -> None:
        self.register_kv_caches({"__cross_layer__": kv_cache})
```

**Review by ivanium — CHANGES_REQUESTED** [review]
Thanks for the PR and thanks @NickLucche for the ping! Overall makes sense but I think we can simplify it. Left a comment.

**Comment by orozery** [comment]
Just FYI this whole `register_cross_layers_kv_cache` flow is dying out as dense models move to MRV2.

**Comment by ivanium** [comment]
> Just FYI this whole `register_cross_layers_kv_cache` flow is dying out as dense models move to MRV2.

Oh thanks for the heads up! I guess we could support cross-layer layout in MRV2 too, but I know Lucas is actively refactoring the layout in general. Curious do you think we will still go with cross-layer layout there?

**Comment by orozery** [comment]
> Curious do you think we will still go with cross-layer layout there?

We should. It has both performance gains for KV transfers, as well as yielding an easier path for TP-agnostic KV cache representation (num_blocks, num_heads, num_tensors, ...)
The KV connector API for would be different though (different than the existing `register_cross_layers_kv_cache`).

**Review by ivanium — APPROVED** [review]
LGTM. Thanks!
